lib/HeroBonus.cpp

```diff
@@ -809,8 +802,16 @@ void CBonusSystemNode::updateBonuses(const CSelector &s)
 		child->updateBonuses(s);
 }
 
-void CBonusSystemNode::addNewBonus(Bonus *b)
+void CBonusSystemNode::addNewBonus(const std::shared_ptr<Bonus>& b)

```

Comment:
No it must be same Bonus objects, there are some places where Bonus is modified after addNewBonus call.
As for smart ptr itself I prefer addNewBonus declaration consistent with removeBonus, and do not care about unique check.